### PR TITLE
Casting Codegen

### LINKED ISSUE #108: Code generation for casting construct
Code generation for ticket #91

Implement code generation of `as!`. This is going to essentially be an in-line inequality check, requiring us to define upper and lower bounds for each of the `SolidityType`s.

### PR DESCRIPTION
**Pull request by nvgrw**
Closes #108

### CODE REVIEW

**Comment by nvgrw** [comment]
Blocked on #129, will add tests once that gets merged

**Comment by nvgrw** [comment]
I'll resume writing some tests once exams are over

**Comment by nvgrw** [comment]
I've written the tests, there are only a few of them.
One problem right now is that I don't know how to pass unsigned values to truffle. Some help on this would be appreciated. For now I suspect it treats the values as signed so I divide by 2 to get the "unsigned" value. Or it's my checking logic.
Opinions welcome.

File: Tests/Integration/BehaviorTests/tests/type_conversions/test/migrations/1.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+var config = require("../config.js")
+
+var Contract = artifacts.require("./" + config.contractName + ".sol");
+var External = artifacts.require("./External.sol");
```
Comment (Aurel300) [inline Tests/Integration/BehaviorTests/tests/type_conversions/test/migrations/1.js]: How come this path is `External.sol`, not `sol_external.sol` ?

File: Tests/Integration/BehaviorTests/tests/type_conversions/test/test/test.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
+// RUN: cd %S && truffle test
+
+var config = require("../config.js")
+
+var Contract = artifacts.require("./" + config.contractName + ".sol");
+var Interface = artifacts.require("./_Interface" + config.contractName + ".sol");
+Contract.abi = Interface.abi
+
+var ExternalContract = artifacts.require("./External.sol");
```
Comment (Aurel300) [inline Tests/Integration/BehaviorTests/tests/type_conversions/test/test/test.js]: Same here.

File: Tests/Integration/BehaviorTests/tests/type_conversions/test/test/test.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    t = await instance.testReduce(100, external.address);
+    assert.equal(t.valueOf(), 100);
+  });
+
+  it("should be possible to cast from a higher bit type to a lower bit type for small values", async function() {
+    const external = await ExternalContract.deployed();
+    const instance = await Contract.deployed();
+
+    let t;
+    const value = 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF / 2;
```
Comment (Aurel300) [inline Tests/Integration/BehaviorTests/tests/type_conversions/test/test/test.js]: No, these cannot work. Anything outside 64-bit is automatically a floating-point number in Javascript.

```
$ node
> 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
3.402823669209385e+38
> 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF - 1
3.402823669209385e+38
> 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF == (0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF - 1)
true
```

Although documentation seems scarce, [this link](https://ethereum.stackexchange.com/questions/29597/how-to-convert-bignumber-to-number-in-truffle-framework) (the question actually) seems to suggest you can use string to represent call arguments with large values.

File: Tests/Integration/BehaviorTests/tests/type_conversions/test/test/test.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    t = await instance.testIncrease(value, external.address);
+    assert.equal(t.valueOf(), value);
+  });
+
+  it("should not be possible to cast from a lower bit type to a higher bit type", async function() {
+    const external = await ExternalContract.deployed();
+    const instance = await Contract.deployed();
+    const value = 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF;
```
Comment (Aurel300) [inline Tests/Integration/BehaviorTests/tests/type_conversions/test/test/test.js]: Likewise.

File: Tests/Integration/BehaviorTests/tests/type_conversions/test/migrations/1.js (hunk 1 above)
Comment (nvgrw) [reply Tests/Integration/BehaviorTests/tests/type_conversions/test/migrations/1.js]: taken from `xcall` directory. It's some odd naming scheme in truffle.

File: Tests/Integration/BehaviorTests/tests/type_conversions/test/test/test.js (hunk 2 above)
Comment (nvgrw) [reply Tests/Integration/BehaviorTests/tests/type_conversions/test/test/test.js]: Same here

File: Tests/Integration/BehaviorTests/tests/type_conversions/test/test/test.js (hunk 3 above)
Comment (nvgrw) [reply Tests/Integration/BehaviorTests/tests/type_conversions/test/test/test.js]: I wasn't sure about this. Thanks so much! The documentation is seriously lacking… I tried finding an API doc for truffle but couldn't even find that. Let's use a string then!

**Pull request merged**